tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1

```diff
@@ -66,38 +146,478 @@ function Get-AzProfile
+  {
+    $Profile = $PSBoundParameters.Profile
+    $Remove = $PSBoundParameters.RemovePreviousVersions
+
+    Do
+    {
+      $moduleInstalled = Get-Module -Name $Module -ListAvailable | Where-Object { $_.Version -eq $version} 
+      if ($PSCmdlet.ShouldProcess("$module version $version", "Remove module")) 
+      {
+        if (($moduleInstalled -ne $null) -and ($Remove.IsPresent -or $PSCmdlet.ShouldContinue("Remove module $Module version $version", "Removing Modules for profile $Profile"))) 
+        {
+          Remove-Module -Name $module -Force -ErrorAction "SilentlyContinue"
+          try 
+          {
+            Uninstall-Module -Name $module -RequiredVersion $version -Force -ErrorAction Stop
+          }
+          catch
+          {
+            Write-Warning $_.Exception.Message
+            break
+          }
+        }
+        else {
+          break
+        }
+      }
+      else {
+        break
+      }
+    }
+    While($moduleInstalled -ne $null);        
+  }
+}
+
+# Help function to uninstall a profile
+function Uninstall-ProfileHelper
+{
+  [CmdletBinding()]
+  param([PSObject]$PMap)
+  DynamicParam
+  {
+    $params = New-Object -Type System.Management.Automation.RuntimeDefinedParameterDictionary
+    Add-ProfileParam $params
+    Add-ForceParam $params
+    return $params
+  }
+
+  PROCESS {
+    $Profile = $PSBoundParameters.Profile
+    $Force = $PSBoundParameters.Force
+    $modules = ($PMap.$Profile | Get-Member -MemberType NoteProperty).Name
+
+    # Get-Profiles installed across all hashes. This is to avoid uninstalling modules that are part of other installed profiles
+    $AllProfilesInstalled = Get-AllProfilesInstalled
+
+    foreach ($module in $modules)
+    {
+      if ($Force.IsPresent)
+      {
+        Invoke-UninstallModule -PMap $PMap -Profile $Profile -Module $module -AllProfilesInstalled $AllProfilesInstalled -RemovePreviousVersions
+      }
+      else {
+        Invoke-UninstallModule -PMap $PMap -Profile $Profile -Module $module -AllProfilesInstalled $AllProfilesInstalled 
+      }
+    }
+  }
+}
+
+# Checks if the module is part of other installed profiles. Calls Uninstall-ModuleHelper if not.
+function Invoke-UninstallModule
+{
+  [CmdletBinding()]
+  param([PSObject]$PMap, [String]$Module, [hashtable]$AllProfilesInstalled)
+  DynamicParam
+  {
+    $params = New-Object -Type System.Management.Automation.RuntimeDefinedParameterDictionary
+    Add-ProfileParam $params
+    Add-RemoveParam $params
+    return $params
+  }
+  
+  PROCESS
+  {
+    $Profile = $PSBoundParameters.Profile
+
+    # Check if the profiles associated with the module version are installed.
+    $profilesAssociated = Test-ProfilesInstalled -Module $Module -Profile $Profile -PMap $PMap -AllProfilesInstalled $AllProfilesInstalled
+      
+    # If more than one profile is installed for the same version of the module, do not uninstall
+    if (($profilesAssociated.Count -gt 1) -and ($profilesAssociated -ne $Profile))
```

Comment:
Makes sense, and if its equal to 1, it always comes from $profile, so no need to check. Removed the second clause